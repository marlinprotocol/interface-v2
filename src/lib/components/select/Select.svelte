<script lang="ts">
	import Text from '$lib/atoms/texts/Text.svelte';
	import CollapseButton from '$lib/components/buttons/CollapseButton.svelte';
	import { cn } from '$lib/utils/helpers/commonHelper';

	export let dataList: (string | number)[] = [];
	export let title = '';
	export let value: string | number | undefined = '';
	export let setValue: ((value: string | number) => any) | undefined = undefined;
	export let showSuggestions = false;
	export let suggestions: (string | number)[] | string[][] = [];
	export let textSuffix = '';
	export let showLabel = false;
	export let disabled = false;

	let windowWidth: number;
	let searchContainer: HTMLDivElement;

	const handleClickOutside = (event: MouseEvent) => {
		if (searchContainer.contains(event.target as Node) === false) {
			showSuggestions = false;
		}
	};

	const handleToggleShowAllSuggestions = () => {
		if (disabled) return;
		suggestions = dataList;
		showSuggestions = !showSuggestions;
	};

	const handleSuggestionClick = async (suggestion: string | number | Array<string>) => {
		showSuggestions = false;
		value =
			typeof suggestion === 'string' || typeof suggestion === 'number' ? suggestion : suggestion[1];
		setValue?.(value);
	};
</script>

<svelte:window on:click={handleClickOutside} bind:innerWidth={windowWidth} />

<!-- svelte-ignore a11y-click-events-have-key-events -->
<!-- svelte-ignore a11y-no-static-element-interactions -->
<div
	data-testid="select"
	id={'select-' + title.toLowerCase()}
	bind:this={searchContainer}
	on:click={handleToggleShowAllSuggestions}
	class={cn('search-container relative flex w-fit items-baseline', {
		'cursor-not-allowed': disabled
	})}
>
	<div class="flex h-auto items-center">
		{#if showLabel}
			<span class="mr-2 text-sm text-primary">{value}</span>
		{/if}
		{#if !disabled}
			<CollapseButton isOpen={showSuggestions} disabled={dataList.length === 0} />
		{/if}
	</div>

	{#if showSuggestions && suggestions.length > 0}
		<ul
			class="shadow-select-dropdown absolute right-[-17px] top-10 z-20 max-h-72 w-fit min-w-[150px] overflow-y-auto rounded-md bg-secondary-content text-base focus:outline-none sm:text-sm"
		>
			<div class="flex items-center justify-between px-8 py-4">
				<Text variant="small" fontWeight="font-semibold" text={title ?? 'Select'} />
			</div>
			{#each suggestions as suggestion}
				<li
					class="cursor-pointer text-left hover:bg-base-200 {Array.isArray(suggestion)
						? suggestion[1] === value
							? 'bg-base-200'
							: 'bg-secondary-content'
						: suggestion === value
							? 'bg-base-200'
							: 'bg-secondary-content'}"
				>
					<button
						class="w-full px-8 py-3 text-left"
						on:click={() => handleSuggestionClick(suggestion)}
					>
						<!-- if the suggestion list has an array/object for displaying suggestions. Eg. region select-->
						{#if Array.isArray(suggestion)}
							<div
								class="flex w-full min-w-[320px] items-baseline justify-between whitespace-nowrap text-sm font-medium"
							>
								<span>{suggestion[0]}</span>
								<span class="text-xs font-normal">{suggestion[1]}</span>
							</div>
						{:else}
							<div class="w-full whitespace-nowrap text-sm font-medium">
								{suggestion + (suggestion !== 'All' ? textSuffix : '')}
							</div>
						{/if}
					</button>
				</li>
			{/each}
		</ul>
	{/if}
</div>

<style>
	/* always show y scrollbar in chrome */
	/* ::-webkit-scrollbar {
		-webkit-appearance: none;
		width: 7px;
	}

	::-webkit-scrollbar-thumb {
		border-radius: 5px;
		background-color: rgba(0, 0, 0, 0.4);
		-webkit-box-shadow: 0 0 1px rgba(255, 255, 255, 0.5);
	}

	::-webkit-scrollbar-track {
		background-color: rgba(0, 0, 0, 0.08);
		border-radius: 5px;
	} */

	::-webkit-scrollbar-track {
		border-radius: 5px;
	}

	::-webkit-scrollbar {
		width: 14px;
	}

	::-webkit-scrollbar-thumb {
		border: 4px solid rgba(0, 0, 0, 0);
		background-clip: padding-box;
		border-radius: 9999px;
		background-color: #aaaaaa;
	}
</style>
