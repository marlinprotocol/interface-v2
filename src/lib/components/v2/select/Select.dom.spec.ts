import { cleanup, fireEvent, render } from '@testing-library/svelte';
import Select from './Select.svelte';

const dataList = ['test1', 'test2', 'test3', 'test4'];

describe('Select', () => {
	afterEach(() => {
		cleanup();
		vi.resetAllMocks();
	});

	it('renders with default props', () => {
		const { getByTestId, container } = render(Select);
		expect(getByTestId('select')).toBeTruthy();
		expect(getByTestId('select').classList.contains('search-container')).toBeTruthy();
		expect(getByTestId('select').contains(getByTestId('collapse-button'))).toBeTruthy();
		expect(container).toMatchSnapshot();
	});

	it('renders with label & datalist props', () => {
		const { getByTestId } = render(Select, { props: { showLabel: true, dataList } });
		expect(getByTestId('select')).toBeTruthy();
		expect(getByTestId('select').contains(getByTestId('collapse-button'))).toBeTruthy();
		expect(getByTestId('select').firstElementChild?.innerHTML).toBeFalsy();
	});

	it('renders with List Toggle', async () => {
		const { getByTestId } = render(Select, { props: { dataList } });
		expect(getByTestId('select')).toBeTruthy();
		expect(getByTestId('select').contains(getByTestId('collapse-button'))).toBeTruthy();
		expect(getByTestId('select').lastElementChild?.classList.contains('absolute')).toBeFalsy();
		await fireEvent.click(getByTestId('select'));
		expect(getByTestId('select').lastElementChild?.classList.contains('absolute')).toBe(true);
		await fireEvent.click(getByTestId('select'));
		expect(getByTestId('select').lastElementChild?.classList.contains('absolute')).toBeFalsy();
	});

	it('fires an event to select item in Select Component when clicked on an item', async () => {
		const { getByTestId } = render(Select, { props: { dataList, showLabel: true } });
		expect(getByTestId('select')).toBeTruthy();
		expect(getByTestId('select').contains(getByTestId('collapse-button'))).toBeTruthy();
		expect(getByTestId('select').lastElementChild?.classList.contains('absolute')).toBeFalsy();
		await fireEvent.click(getByTestId('select'));
		expect(getByTestId('select').lastElementChild?.classList.contains('absolute')).toBe(true);
		const firstListItem = getByTestId('select').lastElementChild?.children[1]
			.firstElementChild as Element;
		await fireEvent.click(firstListItem);
		expect(getByTestId('select').firstElementChild?.innerHTML === 'test1').toBeTruthy();
	});

	it('default open select and use setValue props', async () => {
		let selectedItem = '';
		const { getByTestId } = render(Select, {
			props: {
				dataList,
				showLabel: true,
				showSuggestions: true,
				suggestions: dataList,
				setValue: (value) => {
					selectedItem = value as string;
				}
			}
		});
		expect(getByTestId('select')).toBeTruthy();
		expect(getByTestId('select').contains(getByTestId('collapse-button'))).toBeTruthy();
		expect(getByTestId('select').lastElementChild?.classList.contains('absolute')).toBe(true);
		const firstListItem = getByTestId('select').lastElementChild?.children[1]
			.firstElementChild as Element;
		expect(selectedItem).toBeFalsy();
		await fireEvent.click(firstListItem);
		expect(getByTestId('select').firstElementChild?.innerHTML === selectedItem).toBeTruthy();
	});

	it('closes the select list when clicked outside', async () => {
		const { getByTestId, container } = render(Select, {
			props: {
				dataList,
				showSuggestions: true,
				suggestions: dataList
			}
		});
		expect(getByTestId('select')).toBeTruthy();
		expect(getByTestId('select').contains(getByTestId('collapse-button'))).toBeTruthy();
		expect(getByTestId('select').lastElementChild?.classList.contains('absolute')).toBe(true);
		await fireEvent.click(container);
		expect(getByTestId('select').lastElementChild?.classList.contains('absolute')).toBe(false);
	});
});
