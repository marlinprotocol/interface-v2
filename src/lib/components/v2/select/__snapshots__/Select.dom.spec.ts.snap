// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > renders with default props 1`] = `
<body>
  <div>
    <div
      class="search-container relative flex w-fit cursor-pointer items-baseline s-Q2GksoR5hVxS"
      data-testid="select"
    >
       
      <button
        class="btn btn-xs rounded-full btn-circle p-0 bg-base-300 border-primary/[.2] hover:bg-base-300 disabled:bg-base-200 disabled:cursor-not-allowed btn btn-xs btn-circle p-0 bg-base-300 hover:bg-base-300 border-none disabled:bg-base-200 disabled:cursor-not-allowed"
        data-testid="collapse-button"
        disabled=""
        id=""
      >
        <svg
          aria-label=""
          class="fa-icon icon-disabled s-T5YyQewRBd7n"
          data-testid="icon"
          height="16"
          role="presentation"
          style="width:12px; height:12px; cursor:pointer;  "
          version="1.1"
          viewBox="0 0 1792 1792"
          width="16"
        >
          <path
            d="M1683 808l-742 741q-19 19-45 19t-45-19l-742-741q-19-19-19-45.5t19-45.5l166-165q19-19 45-19t45 19l531 531 531-531q19-19 45-19t45 19l166 165q19 19 19 45.5t-19 45.5z"
          />
          
           
           
          
          
        </svg>
        <!--&lt;Svg&gt;-->
        <!--&lt;Icon&gt;-->
        <!--&lt;Icon&gt;-->
      </button>
      <!--&lt;CollapseButton&gt;-->
       
    </div>
    <!--&lt;Select&gt;-->
  </div>
</body>
`;
